server/channels/store/sqlstore/channel_store_categories.go

```diff
@@ -748,49 +752,67 @@ func (s SqlChannelStore) UpdateSidebarCategories(userId, teamId string, categori
 		if _, err = transaction.Exec(updateQuery, updateParams...); err != nil {
 			return nil, nil, errors.Wrap(err, "failed to update SidebarCategories")
 		}
+	}
+
+	// Second, update the channels in those categories
+	categoryIds := make([]string, len(categories))
+	for i, category := range categories {
+		categoryIds[i] = category.Id
+	}
+
+	// Remove any SidebarChannels entries that were previously in this category. This needs to be done for all
+	// categories at once to prevent deadlocks.
+	//
+	// Note that this means that moving channels between categories requires updating both the source and
+	// destination categories.
+	query, args, err2 := s.getQueryBuilder().
+		Delete("SidebarChannels").
+		Where(sq.Eq{"CategoryId": categoryIds}).ToSql()
+	if err2 != nil {
+		return nil, nil, errors.Wrap(err2, "update_sidebar_categories_tosql2")
+	}
+
+	if _, err = transaction.Exec(query, args...); err != nil {
+		return nil, nil, errors.Wrap(err, "failed to delete SidebarChannels")
+	}
+
+	for _, category := range categories {
+		if category.Type == model.SidebarCategoryDirectMessages {
+			// The order of the DM category isn't stored explicitly, so there's nothing to do here
+			continue
+		}
 
-		// if we are updating DM category, it's order can't channel order cannot be changed.
-		if category.Type != model.SidebarCategoryDirectMessages {
-			// Remove any SidebarChannels entries that were either:
-			// - previously in this category (and any ones that are still in the category will be recreated below)
-			// - in another category and are being added to this category
-			query, args, err2 := s.getQueryBuilder().
-				Delete("SidebarChannels").
-				Where(
-					sq.And{
-						sq.Eq{"ChannelId": srcCategory.Channels},
-						sq.Eq{"CategoryId": category.Id},
-					},
-				).ToSql()
+		runningOrder := 0
+		insertQuery := s.getQueryBuilder().
+			Insert("SidebarChannels").
+			Columns("ChannelId", "UserId", "CategoryId", "SortOrder")
 
-			if err2 != nil {
-				return nil, nil, errors.Wrap(err2, "update_sidebar_categories_tosql2")
-			}
+		if s.DriverName() == model.DatabaseDriverMysql {
+			insertQuery = insertQuery.Suffix("ON DUPLICATE KEY UPDATE SortOrder = VALUES(SortOrder)")
+		} else {
+			insertQuery = insertQuery.Suffix("ON CONFLICT (ChannelId, UserId, CategoryId) DO UPDATE SET SortOrder = excluded.SortOrder")
```

Comment:
Postgres actually has the option to use `DO NOTHING`. Let's do that here like we do in other places.

Reply:
I'd like to leave this because I want the last update's sort order to win. Is that okay?

Reply:
Oh I see. Sure.